### ISSUE TITLE
mysterious empty space at the end of the bar

### ISSUE THREAD

**Issue by StoicIndividual** [bakkeby/dwm-flexipatch#370]
Hi i've been trying to configure status2d and noticed extra spacing at the end of the bar. 

Now this normally wouldn't be an issue but i found that it also removes any space i add to a module, like printf `"^b#C96D2A^^c#B8A598^hello "` or `echo -e "^b#C96D2A^^c#B8A598^hello "`
![image](https://github.com/bakkeby/dwm-flexipatch/assets/44035542/f59a193c-5542-4357-9b32-7896c1f884fb)

the space was also added when i tried to patch luke smith's dwm with statuspadding patch, but that patch is disabled on the current config with dwm flexipatch  but i did enable it before, i'm not sure if it some of it's code became persistant even after removing it

**Comment by bakkeby** [comment]
Do you have systray enabled?

Also could you list the patches you have enabled in case it is something else?
```sh
grep -E "^#define .* 1$" patches.h
```

**Comment by StoicIndividual** [comment]
No i don't have systray enabled, also here are the patches:

```
#define BAR_DWMBLOCKS_PATCH 1
#define BAR_LTSYMBOL_PATCH 1
#define BAR_STATUS_PATCH 1
#define BAR_STATUSCMD_PATCH 1
#define BAR_STATUS2D_PATCH 1
#define BAR_STATUS2D_XRDB_TERMCOLORS_PATCH 1
#define BAR_TAGS_PATCH 1
#define BAR_UNDERLINETAGS_PATCH 1
#define BAR_WINICON_PATCH 1
#define BAR_WINTITLE_PATCH 1
#define BAR_CENTEREDWINDOWNAME_PATCH 1
#define BAR_HIDEVACANTTAGS_PATCH 1
#define BAR_PADDING_VANITYGAPS_PATCH 1
#define NOBORDER_PATCH 1
#define RESTARTSIG_PATCH 1
#define SEAMLESS_RESTART_PATCH 1
#define VANITYGAPS_PATCH 1
#define XRDB_PATCH 1
#define ZOOMFLOATING_PATCH 1
#define TILE_LAYOUT 1
#define MONOCLE_LAYOUT 1
```

**Comment by StoicIndividual** [comment]
Ah I should probably mention that it's dwmblocks async that i'm using

**Comment by bakkeby** [comment]
You should have the same empty space at the start of the status as well; it is just the left / right padding.

This happens for normal status as well not just status2d. It is just more noticeable when you draw the background colour like that.

The padding is intentional so that the text is not crammed up against the edge of the bar. In any case it is not something that is configurable.

If you wanted to remove the padding for a normal status then this change should work.

```diff
diff --git a/patch/bar_status.c b/patch/bar_status.c
index 715b6d2..1910172 100644
--- a/patch/bar_status.c
+++ b/patch/bar_status.c
@@ -1,7 +1,7 @@
 int
 width_status(Bar *bar, BarArg *a)
 {
-       return TEXTWM(stext);
+       return TEXTWM(stext) - lrpad;
 }

 #if BAR_EXTRASTATUS_PATCH
@@ -15,7 +15,7 @@ width_status_es(Bar *bar, BarArg *a)
 int
 draw_status(Bar *bar, BarArg *a)
 {
-       return drw_text(drw, a->x, a->y, a->w, a->h, lrpad / 2, stext, 0, True);
+       return drw_text(drw, a->x, a->y, a->w, a->h, 0, stext, 0, True);
 }

 #if BAR_EXTRASTATUS_PATCH
```

and for the status2d you would do.

```diff
diff --git a/patch/bar_status2d.c b/patch/bar_status2d.c
index 860fd83..00e4139 100644
--- a/patch/bar_status2d.c
+++ b/patch/bar_status2d.c
@@ -30,7 +30,7 @@ width_status2d(Bar *bar, BarArg *a)
        #else
        width = status2dtextlength(stext);
        #endif // #if BAR_EXTRASTATUS_PATCH | BAR_STATUSCMD_PATCH
-       return width ? width + lrpad : 0;
+       return width ? width : 0;
 }

 #if BAR_EXTRASTATUS_PATCH
@@ -97,7 +97,6 @@ drawstatusbar(BarArg *a, char* stext)
        memcpy(text, stext, len);
        #endif // BAR_STATUSCMD_PATCH

-       x += lrpad / 2;
        drw_setscheme(drw, scheme[LENGTH(colors)]);
        drw->scheme[ColFg] = scheme[SchemeNorm][ColFg];
        drw->scheme[ColBg] = scheme[SchemeNorm][ColBg];
```

**Comment by StoicIndividual** [comment]
i made the changes in the file and ran `sudo make clean install` and restarted the session, but the space still persists, also the space isn't that bad it's just i can't seem to add any space to the right side of the orange wifi module
![image](https://github.com/bakkeby/dwm-flexipatch/assets/44035542/aa7e8a74-74b8-4bd2-8d7c-7214eba2faa0)
because for some reason something is essentially "removing" the space character from the printf command (`printf "^b#C96D2A^^c#B8A598^hello <this space isn't registered>`" and i just assumed that the extra padding at the end of the bar had something to do with it but I could be wrong.

this is the script i'm using if it helps:

#!/bin/sh

if grep -xq 'up' /sys/class/net/w*/operstate 2>/dev/null ; then
	wifiicon="$(awk '/^\s*w/ { print "󰒢", int($3 * 100 / 70) "% " }' /proc/net/wireless)"
elif grep -xq 'down' /sys/class/net/w*/operstate 2>/dev/null ; then
	grep -xq '0x1003' /sys/class/net/w*/flags && wifiicon="󰤉  " || wifiicon="󰞃 "
fi

printf "^b#C96D2A^^c#B8A598^ %s%s%s\n" "$wifiicon" "$(sed "s/down/󰈂 /;s/up/󰈁/" /sys/class/net/e*/operstate 2>/dev/null)" "$(sed "s/.*/🔒/" /sys/class/net/tun*/operstate 2>/dev/null)"

**Comment by bakkeby** [comment]
When you say you made changes in the file, which file are you referring to?

For the script you could try adding the extra space in the printf format. Not sure if dwmblocks will remove trailing spaces, if that is the case you could end the line with `^d^`.

**Comment by bakkeby** [comment]
Another thing you can try is to run your script directly via the command line.

```sh
xsetroot -name "$(./yourscript.sh)"
```
and see if that makes a difference.

I'm only mentioning this because dwmblocks has a limit on how many characters a single block can have. In dwmblocks-async this limit has been moved to [config.h](https://github.com/UtkarshVerma/dwmblocks-async/blob/main/config.h#L4C9-L4C18).

**Comment by StoicIndividual** [comment]
good news the adding ^d^ to the end of the printf format works like:
`printf "^b#C96D2A^^c#B8A598^ %s%s%s^d^\n" "$wifiicon" "$(sed "s/down/󰈂 /;s/up/󰈁/" /sys/class/net/e*/operstate 2>/dev/null)" "$(sed "s/.*/🔒/" /sys/class/net/tun*/operstate 2>/dev/null)"
`
![image](https://github.com/bakkeby/dwm-flexipatch/assets/44035542/5a645953-b333-47d2-a9a4-11da6f745791)
I still don't know why it works since status2d says that ^d^ is for "Reset colors to SchemeNorm." and i assumed it had something to to with resetting colors.
Also the files i made changes to remove the padding are /dwm-flexipatch/patch/bar_status.c and dwm-flexipatch/patch/bar_status2d.c

**Comment by bakkeby** [comment]
Yes ^d^ is to reset colors back to default, but it works because your line does not end with one or more spaces. 

The dwmblocks-async trims trailing spaces by the looks of it. 

https://github.com/UtkarshVerma/dwmblocks-async/blob/27731295335d31966491f55f54912c3b9dd1a7a7/src/block.c#L44

https://github.com/UtkarshVerma/dwmblocks-async/blob/27731295335d31966491f55f54912c3b9dd1a7a7/src/util.c#L37

**Comment by StoicIndividual** [comment]
Oh I sent the script without the space added my bad,  but adding any amount of space characters like `printf "^b#C96D2A^^c#B8A598^ %s%s%s      \n" "$wifiicon" "$(sed "s/down/󰈂 /;s/up/󰈁/" /sys/class/net/e*/operstate 2>/dev/null)" "$(sed "s/.*/🔒/" /sys/class/net/tun*/operstate 2>/dev/null)" ` woudn't work since the space was constantly being removed by something, which now appears to be dwmblocks, but only after adding ^d^ at the end of the printf format does it add extra space which is weird
